internal/resource/tree.go

```diff
@@ -0,0 +1,164 @@
+package resource
+
+import (
+	"encoding/json"
+	"slices"
+
+	apiv1 "github.com/Azure/eno/api/v1"
+	"github.com/emirpasic/gods/v2/trees/redblacktree"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+type indexedResource struct {
+	Resource            *Resource
+	State               *apiv1.ResourceState
+	PendingDependencies map[Ref]struct{}
+	Dependents          map[Ref]*indexedResource
+}
+
+// treeBuilder is used to index a set of resources into a stateTree.
+type treeBuilder struct {
+	byRef        map[Ref]*indexedResource
+	byGroup      *redblacktree.Tree[int, []*indexedResource]
+	byDefiningGK map[schema.GroupKind]*indexedResource
+	byGK         map[schema.GroupKind]*indexedResource
+}
+
+func (b *treeBuilder) Add(resource *Resource) {
+	// Initialize the builder
+	if b.byRef == nil {
+		b.byRef = map[Ref]*indexedResource{}
+	}
+	if b.byGroup == nil {
+		b.byGroup = redblacktree.New[int, []*indexedResource]()
+	}
+	if b.byDefiningGK == nil {
+		b.byDefiningGK = map[schema.GroupKind]*indexedResource{}
+	}
+	if b.byGK == nil {
+		b.byGK = map[schema.GroupKind]*indexedResource{}
+	}
+
+	// Handle conflicting refs deterministically
+	if existing, ok := b.byRef[resource.Ref]; ok && resource.Less(existing.Resource) {
+		return
+	}
+
+	// Index the resource into the builder
+	idx := &indexedResource{
+		Resource:            resource,
+		PendingDependencies: map[Ref]struct{}{},
+		Dependents:          map[Ref]*indexedResource{},
+	}
+	b.byRef[resource.Ref] = idx
+	current, _ := b.byGroup.Get(resource.ReadinessGroup)
+	b.byGroup.Put(resource.ReadinessGroup, append(current, idx))
+	b.byGK[resource.GVK.GroupKind()] = idx
+	if resource.DefinedGroupKind != nil {
+		b.byDefiningGK[*resource.DefinedGroupKind] = idx
+	}
+}
+
+func (b *treeBuilder) Build() *tree {
+	t := &tree{
+		byRef:     b.byRef,
+		byManiRef: map[ManifestRef]*indexedResource{},
+	}
+
+	for _, idx := range b.byRef {
+		t.byManiRef[idx.Resource.ManifestRef] = idx
+
+		// CRs are dependent on their CRDs
+		i := b.byGroup.IteratorAt(b.byGroup.GetNode(idx.Resource.ReadinessGroup))
+		crd, ok := b.byDefiningGK[idx.Resource.GVK.GroupKind()]
+		if ok {
+			idx.PendingDependencies[crd.Resource.Ref] = struct{}{}
+			crd.Dependents[idx.Resource.Ref] = idx
+		}
+
+		// Depend on any resources in the previous readiness group
+		if i.Prev() {
```

Comment:
I have PTSD from the last bug, Prev and Next don't rely on adjacency right?

Reply:
lolol Right - the iterator behaves correctly in this case. Just hopping between the nodes using their pointers is where we got into trouble before.